lib/nm/index.ts

```diff
@@ -0,0 +1,377 @@
+		if (_.isUndefined(ap)) {
+			return Bluebird.reject(formatError(404)(`Could not find neaby AccessPoints with SSID: ${network.ssid}`)());
+		}
+
+		const [listConnectionsErr, results] = await to(this.listConnections());
+		if (listConnectionsErr) {
+			throw formatError(500)(`Could not listConnections`)(listConnectionsErr);
+		}
+
+		const wifiConnection = findConnection(results)(network);
+		// Check if network's already registered on the device
+		if (!_.isUndefined(wifiConnection)) {
+			const [activateConnectionErr] = await to(this.activateConnection([wifiConnection.path, this.devices.wifi.path, '/']));
+			if (activateConnectionErr) {
+				throw formatError(500)(`Could not activateConnection`)(activateConnectionErr);
+			}
+			return;
+		}
+
+		const [addConnectionErr, networkSettings] = await to(this.addConnection(connectionParam));
+		if (addConnectionErr) {
+			throw formatError(500)(`Could not addConnection`)(addConnectionErr);
+		}
+
+		const [activateConnectionErr] = await to(this.activateConnection([networkSettings, this.devices.wifi.path, '/']));
+		if (activateConnectionErr) {
+			throw formatError(500)(`Could not activateConnection`)(activateConnectionErr);
+		}
+		return;
+	}
+
+	listConnections = async () => {
+		const [ListConnectionsErr, connections] = await to(this.callMethod(['Settings'])(['Settings'])(['ListConnections', ''])());
+		if (ListConnectionsErr) {
+			throw formatError(500)(`Could not ListConnections`)(ListConnectionsErr);
+		}
+
+		const getConnectionsSettings = () => _.map(connections, this.getConnectionSettings);
+		const [getConnectionsSettingsErr, results] = await to(Bluebird.all(getConnectionsSettings()));
+		if (getConnectionsSettingsErr) {
+			throw formatError(500)(`Could not getConnectionsSettings`)(getConnectionsSettingsErr);
+		}
+		return results;
+	}
+
+	forgetNetwork = async (network) => {
+		const [listConnectionsErr, results] = await to(this.listConnections());
+		if (listConnectionsErr) {
+			throw formatError(500)(`Could not listConnections`)(listConnectionsErr);
+		}
+		const wifiConnection = findConnection(results)(network);
+
+		const [deleteConnectionErr] = await to(this.deleteConnection(wifiConnection.path));
+		if (deleteConnectionErr) {
+			throw formatError(500)(`Could not deleteConnection`)(deleteConnectionErr);
+		}
+		return Bluebird.resolve();
+	}
+
+	listNearbyNetworks = async () => {
+		const requestScanParams = [
+			['ssids', ['aay', [stringToArrayOfBytes('1')]]],
+		];
+		const [requestScanErr] = await to(this.requestScan(requestScanParams));
+		if (requestScanErr) {
+			throw formatError(500)(`Could not requestScan`)(requestScanErr);
+		}
+		const [getDeviceAccessPointsErr, {AccessPoints}] = await to(this.getObjectProperty(['org.freedesktop.NetworkManager.Device.Wireless','AccessPoints'])(this.devices.wifi.path));
+		if (getDeviceAccessPointsErr) {
+			throw formatError(500)(`Could not get AccessPoints property on ${this.devices.wifi.path}`)(getDeviceAccessPointsErr);
+		}
```

Comment:
I don't get the point of this `to` function.
This is equivalent to the following (which is understandable and doesn't require you to go to the definition of `to`)
```typescript
try {
    { AccessPoints } = await this.getObjectProperty(['org.freedesktop.NetworkManager.Device.Wireless','AccessPoints'])(this.devices.wifi.path)
}catch (error) {
    throw formatError(500)(`Could not get AccessPoints property on ${this.devices.wifi.path}`)(error);
}
```

Reply:
Also `formatError` returning a function returning a function returning a function is weird. Can't we make this simpler?
Like a function with 3 arguments for example?

Reply:
The point was to actually avoid `try...catch`, since it is more expensive than having that `to` function, but if it improves readability for others, I'm up to including `try...catch`.
Gonna switch to it in the next commit batch